src/main/java/com/google/devtools/build/lib/cmdline/Label.java

```diff
@@ -203,6 +211,31 @@ public static Label parseWithPackageContext(String raw, PackageContext packageCo
     return createUnvalidated(PackageIdentifier.create(repoName, pkgFragment), parts.target());
   }
 
+  public static final class RepoMappingRecorder {
+    /** {@code <fromRepo, apparentRepoName, canonicalRepoName> } */
+    Table<RepositoryName, String, RepositoryName> entries = HashBasedTable.create();
+
+    public ImmutableTable<RepositoryName, String, RepositoryName> recordedEntries() {
+      return ImmutableTable.<RepositoryName, String, RepositoryName>builder()
+          .orderRowsBy(Comparator.comparing(RepositoryName::getName))
+          .orderColumnsBy(Comparator.naturalOrder())
+          .putAll(entries)
+          .buildOrThrow();
+    }
+  }
+
+  public static Label parseWithPackageContextRecordingRepoMapping(String raw,
```

Comment:
The extra `RecordingRepoMapping` in the method name is redundant: it's clear that the repo mapping is being recorded because a `RepoMappingRecorder` is being passed in. In general, leave method names short and trust the compiler to handle overloads (or tell you that it can't, and then add suffixes to disambiguate).

The same applies to `parseWithPackageContextInternal` (and, possibly, to `parseWithPackageContext` vs `parse`, but that's older than this PR).

Reply:
fair point about `parseWithPackageContextRecordingRepoMapping`; the others though I'd argue `fooInternal` is useful to clearly denote that a method is an internal implementation detail and not just different from the public `foo` because of an accidental param type difference. Same with `parseWithPackageContext` vs `parse`; the semantics are clearly different, not to mention that `PackageContext` and `RepoContext` are both interfaces, so callers can easily be confused by the overload.